### PR TITLE
Reduce colrefs by skipping unused ones

### PR DESCRIPTION
**Pull request by asubramanya**
Currently we retain all the colrefs of a table even though we only reference a
few of them in the original query. This patch marks the colrefs as used if
either its a system column or if its referenced in the query and unused
otherwise. When we derive the relational properties, we exclude the unused
columns from the output columns list. For a SELECT COUNT(*) query on a 1600
column table which has a 10 way join, we see an improvement of approximately
40%.

Co-authored-by: Abhijit Subramanya <asubramanya@pivotal.io>

### CODE REVIEW

File: libgpopt/include/gpopt/translate/CTranslatorDXLToExpr.h [hunk 1]
```diff
@@ -398,6 +398,9 @@ namespace gpopt
 			CColRef *LookupColRef(UlongToColRefMap *colref_mapping, ULONG colid);
 
 		public:
+
+			static BOOL translating;
```
Comment (zellerh) [inline libgpopt/include/gpopt/translate/CTranslatorDXLToExpr.h]: The public member works just fine, but to be more conventional, maybe make this a private member and define methods to set/reset this?

File: libgpopt/src/operators/CLogicalGet.cpp [hunk 2]
```diff
@@ -231,7 +232,13 @@ CLogicalGet::PcrsDeriveOutput
 	)
 {
 	CColRefSet *pcrs = GPOS_NEW(mp) CColRefSet(mp);
-	pcrs->Include(m_pdrgpcrOutput);
+	for (ULONG i = 0; i < m_pdrgpcrOutput->Size(); i++)
+	{
+		if ((*m_pdrgpcrOutput)[i]->IsUsed() || CTranslatorDXLToExpr::translating)
```
Comment (zellerh) [inline libgpopt/src/operators/CLogicalGet.cpp]: We should add a comment here why we have this extra check.

File: server/src/unittest/gpopt/operators/CExpressionTest.cpp [hunk 3]
```diff
@@ -180,6 +181,28 @@ CExpressionTest::EresUnittest_SimpleOps()
...
+		CLogicalGet *popGet = dynamic_cast<CLogicalGet *>(pexpr->Pop());
+		CLogicalDynamicGetBase *popDynGet = dynamic_cast<CLogicalDynamicGetBase *>(pexpr->Pop());
+		CColRefArray *colrefs = NULL;
+
+		if (NULL != popGet)
+		{
+			colrefs = popGet->PdrgpcrOutput();
+		}
+
+		if (NULL != popDynGet)
```
Comment (zellerh) [inline server/src/unittest/gpopt/operators/CExpressionTest.cpp]: Use an else if here?

File: data/dxl/minidump/AnySubq-With-NonScalarSubqueryChild-1.mdp [hunk 4]
```diff
@@ -541,48 +541,40 @@
...
                 <dxl:OneTimeFilter/>
                 <dxl:Aggregate AggregationStrategy="Sorted" StreamSafe="false">
                   <dxl:Properties>
-                    <dxl:Cost StartupCost="0" TotalCost="1356691870.328568" Rows="1.800000" Width="1"/>
+                    <dxl:Cost StartupCost="0" TotalCost="1356691678.834464" Rows="1.800000" Width="1"/>
                   </dxl:Properties>
                   <dxl:GroupingColumns>
-                    <dxl:GroupingColumn ColId="10"/>
                     <dxl:GroupingColumn ColId="11"/>
-                    <dxl:GroupingColumn ColId="12"/>
```
Comment (bhuvnesh2703) [inline data/dxl/minidump/AnySubq-With-NonScalarSubqueryChild-1.mdp]: 2 grouping cols removed? could you explain why?

File: data/dxl/minidump/AssertMaxOneRow.mdp [hunk 5]
```diff
@@ -1783,42 +1725,14 @@ where oid= (select reltoastrelid from pg_class where relname='fsts_alter_set_sto
                     <dxl:IndexCondList>
                       <dxl:Comparison ComparisonOperator="=" OperatorMdid="0.93.1.0">
                         <dxl:Ident ColId="76" ColName="relname" TypeMdid="0.19.1.0"/>
-                        <dxl:ConstValue TypeMdid="0.19.1.0" IsNull="false" IsByValue="false" Value="ZnN0c19hbHRlcl9zZXRfc3RvcmFnZV90YWJsZQAAAAAAAAAAAAAAAAAAAAAA&#10;AAAAAAAAAAAAAAAAAAAAAAAAAA=="/>
+                        <dxl:ConstValue TypeMdid="0.19.1.0" IsNull="false" IsByValue="false" Value="ZnN0c19hbHRlcl9zZXRfc3RvcmFnZV90YWJsZQAAAAAAAAAAAAAAAAAAAAAA&#xA;AAAAAAAAAAAAAAAAAAAAAAAAAA=="/>
```
Comment (bhuvnesh2703) [inline data/dxl/minidump/AssertMaxOneRow.mdp]: why the change from `#10` => `xA`

File: data/dxl/minidump/CPhysicalParallelUnionAllTest/ParallelUnionAllWithNoRedistributableColumns.mdp [hunk 6]
```diff
@@ -246,130 +246,58 @@ EXPLAIN SELECT xmin FROM tbl1 UNION ALL SELECT xmin FROM tbl2;
...
+            <dxl:Cost StartupCost="0" TotalCost="862.000019" Rows="1.000000" Width="4"/>
           </dxl:Properties>
           <dxl:ProjList>
             <dxl:ProjElem ColId="2" Alias="xmin">
               <dxl:Ident ColId="2" ColName="xmin" TypeMdid="0.28.1.0"/>
             </dxl:ProjElem>
           </dxl:ProjList>
           <dxl:Filter/>
-          <dxl:Result>
+          <dxl:TableScan>
```
Comment (bhuvnesh2703) [inline data/dxl/minidump/CPhysicalParallelUnionAllTest/ParallelUnionAllWithNoRedistributableColumns.mdp]: this test is for parallel union all, so we should see a redistribute motion (no-op) on top of every child of append? why its removed?

File: libgpopt/include/gpopt/base/CUtils.h [hunk 7]
```diff
@@ -1194,7 +1194,6 @@ namespace gpopt
 	{
 		GPOS_ASSERT(NULL != pexprLeft);
 		GPOS_ASSERT(NULL != pexprRight);
-		GPOS_ASSERT(NULL != pcrInner);
```
Comment (bhuvnesh2703) [inline libgpopt/include/gpopt/base/CUtils.h]: we expect inner to be NULL now sometimes, why?

File: libgpopt/src/base/CColumnFactory.cpp [hunk 8]
```diff
@@ -262,12 +262,18 @@ CColumnFactory::PcrCreate
...
 	const CName &name,
-	ULONG ulOpSource
+	ULONG ulOpSource,
+	BOOL mark_as_used
 	)
 {
 	ULONG id = m_aul.Incr();
 	
-	return PcrCreate(pcoldesc, id, name, ulOpSource);
+	CColRef *colref = PcrCreate(pcoldesc, id, name, ulOpSource);
```
Comment (bhuvnesh2703) [inline libgpopt/src/base/CColumnFactory.cpp]: minor: what if we pass mark_as_used as parameter to PcrCreate?

File: libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 9]
```diff
@@ -7302,20 +7302,24 @@ CTranslatorExprToDXL::MakeDXLTableDescr
...
 		}
 		else
 		{
 			colref = m_pcf->PcrCreate(pcd->RetrieveType(), pcd->TypeModifier(), pcd->Name());
 		}
 
-		// use the col ref id for the corresponding output output column as 
+		CMDName *pmdnameCol = GPOS_NEW(m_mp) CMDName(m_mp, pcd->Name().Pstr());
+
+		// use the col ref id for the corresponding output output column as
```
Comment (bhuvnesh2703) [inline libgpopt/src/translate/CTranslatorExprToDXL.cpp]: `output output`

**Comment by bhuvnesh2703** [comment]
Few questions:
a) Instead of true / false for identifying if a column is used or not, could we have an enum which has the following states (chose any name)
1. Default State
2. Used
3. Unused

You can change Default State to Unused, but if something is marked Used we can't change it to Unused.
Do you forsee issue where we have something marked as used (really), but for some reason we mark it unused later because we don't look it up?

b) How we handle computed columns since they will not be originally in the query? Since we initialize starting with true, is it fine because of that?

File: libgpopt/src/optimizer/COptimizer.cpp [hunk 10]
```diff
@@ -243,8 +243,10 @@ COptimizer::PdxlnOptimize
 
 			// translate DXL Tree -> Expr Tree
 			CTranslatorDXLToExpr dxltr(mp, md_accessor);
+			CTranslatorDXLToExpr::translating = true;
```
Comment (sambitesh) [inline libgpopt/src/optimizer/COptimizer.cpp]: Talked to @zellerh about this. It will be good to have comment added here.

File: server/src/unittest/gpopt/translate/CTranslatorDXLToExprTest.cpp [hunk 11]
```diff
@@ -118,14 +118,14 @@ CTranslatorDXLToExprTest::Pexpr
 
 	// translate DXL Tree -> Expr Tree
 	CTranslatorDXLToExpr *pdxltr = GPOS_NEW(mp) CTranslatorDXLToExpr(mp, md_accessor);
-
+	pdxltr->translating = true;
```
Comment (hardikar) [inline server/src/unittest/gpopt/translate/CTranslatorDXLToExprTest.cpp]: Please move this inside PexprTranslateQuery. You can have it as a private variable then. I'm not sure why this is needed at all? Aren't you by-default translating if you're in the PexprTranslateQuery() function?

File: libgpopt/include/gpopt/translate/CTranslatorDXLToExpr.h (hunk 1 above)
Comment (hardikar) [reply libgpopt/include/gpopt/translate/CTranslatorDXLToExpr.h]: Why does this need to be static when the class is used non-static-ly. That is, we create an object of CTranslatorDXLToExpr when we need to use it. Using a static boolean variable to keep track of this could create trouble if there are multiple translations happening at the same time (doesn't happen right now). Consider rewriting this as a member variable.

File: libgpopt/src/translate/CTranslatorDXLToExprUtils.cpp [hunk 12]
```diff
@@ -258,7 +258,8 @@ CTranslatorDXLToExprUtils::Pdrgpcr
 	for (ULONG ul = 0; ul < colids->Size(); ul++)
 	{
 		ULONG *pulColId = (*colids)[ul];
-		const CColRef *colref = colref_mapping->Find(pulColId);
+		CColRef *colref = colref_mapping->Find(pulColId);
```
Comment (hardikar) [inline libgpopt/src/translate/CTranslatorDXLToExprUtils.cpp]: Maybe replace these two lines with `LookupColRef(colref_mapping, pulColId) so that MarkAsUsed gets called automatically.

File: libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 13]
```diff
@@ -7302,20 +7302,24 @@ CTranslatorExprToDXL::MakeDXLTableDescr
...
 		GPOS_ASSERT(NULL != pcd);
 
-		CMDName *pmdnameCol = GPOS_NEW(m_mp) CMDName(m_mp, pcd->Name().Pstr());
-
 		// output col ref for the current col descrs
 		CColRef *colref = NULL;
 		if (NULL != pdrgpcrOutput)
 		{
 			colref = (*pdrgpcrOutput)[ul];
+			if (!colref->IsUsed())
```
Comment (hardikar) [inline libgpopt/src/translate/CTranslatorExprToDXL.cpp]:
```suggestion
			// Don't send unused colrefs to improve performance
			if (!colref->IsUsed())
```

File: server/src/unittest/gpopt/operators/CExpressionTest.cpp (hunk 3 above)
Comment (hardikar) [reply server/src/unittest/gpopt/operators/CExpressionTest.cpp]: Hmm.. I'd prefer to use an if case here comparing `pexpr->Pop()->Eopid()`, instead of dynamic_casts

**Comment by hardikar** [comment]
Ok i just saw @bhuvnesh2703 's commit, it seems he has the same suggestion(s).

File: server/src/unittest/gpopt/minidump/CMiniDumperDXLTest.cpp [hunk 14]
```diff
@@ -151,7 +151,6 @@ CMiniDumperDXLTest::EresUnittest_Basic()
 													ptroutput->GetOutputColumnsDXLArray(),
 													ptroutput->GetCTEProducerDXLArray()
 													);
-
```
Comment (ashuka24) [inline server/src/unittest/gpopt/minidump/CMiniDumperDXLTest.cpp]: this file change seems unnecessary

File: server/src/unittest/CTestUtils.cpp [hunk 15]
```diff
@@ -3095,7 +3106,6 @@ CTestUtils::PexprReadQuery
 						ptroutput->GetOutputColumnsDXLArray(),
 						ptroutput->GetCTEProducerDXLArray()
 						);
-
```
Comment (ashuka24) [inline server/src/unittest/CTestUtils.cpp]: unnecessary change

File: server/src/unittest/gpopt/translate/CTranslatorDXLToExprTest.cpp [hunk 16]
```diff
@@ -118,14 +118,12 @@ CTranslatorDXLToExprTest::Pexpr
 
 	// translate DXL Tree -> Expr Tree
 	CTranslatorDXLToExpr *pdxltr = GPOS_NEW(mp) CTranslatorDXLToExpr(mp, md_accessor);
-
```
Comment (ashuka24) [inline server/src/unittest/gpopt/translate/CTranslatorDXLToExprTest.cpp]: seems like unnecessary file change

**Comment by asubramanya** [comment]
@hardikar @bhuvnesh2703 I changed the implementation to use the enums as you suggested and got rid of the translating flag. Also added a traceflag to enable/disable the feature.

File: libgpopt/include/gpopt/base/CUtils.h (hunk 7 above)
Comment (hardikar) [reply libgpopt/include/gpopt/base/CUtils.h]: @asubramanya Was this concern addressed?

File: libnaucrates/include/naucrates/traceflags/traceflags.h [hunk 17]
```diff
@@ -177,6 +177,10 @@ namespace gpos
 
 		// Eager Agg 
 		EopttraceEnableEagerAgg = 103030,
+		
+		// Enable reduction of unused colrefs
+		EopttraceTranslateUnusedColrefs = 103031,
```
[the remark this answers is not in the document]
Comment (hardikar) [reply libnaucrates/include/naucrates/traceflags/traceflags.h]: // Disable the traceflag to remove unused colrefs

File: libgpopt/src/operators/CLogical.cpp [hunk 18]
```diff
@@ -756,7 +756,7 @@ CLogical::PpcDeriveConstraintFromTable
 
 		pdrgpcrNonSystem->Append(colref);
 
-		if (pcoldesc->IsNullable())
+		if (pcoldesc->IsNullable() || !colref->IsUsed())
```
Comment (hardikar) [inline libgpopt/src/operators/CLogical.cpp]: Should this be `!colref->IsUsed()` or `!colref->IsUnsed()`?

File: libgpopt/src/operators/CLogicalGet.cpp [hunk 19]
```diff
@@ -231,7 +232,17 @@ CLogicalGet::PcrsDeriveOutput
...
 {
 	CColRefSet *pcrs = GPOS_NEW(mp) CColRefSet(mp);
-	pcrs->Include(m_pdrgpcrOutput);
+	for (ULONG i = 0; i < m_pdrgpcrOutput->Size(); i++)
+	{
+		// We want to limit the output columns to only those which are referenced in the query
+		// We will know the entire list of columns which are referenced in the query only after
+		// translating the entire DXL to an expression. Hence we should not limit the output columns
+		// before we have processed the entire DXL.
+		if ((*m_pdrgpcrOutput)[i]->IsUsed() || (*m_pdrgpcrOutput)[i]->IsUnknown())
```
Comment (hardikar) [inline libgpopt/src/operators/CLogicalGet.cpp]: Alternatively, you could use colref->GetUsage() and return the enum. Return Used if it's a system column (also probably a good idea to set it correctly in the constructor). Honestly I'd much prefer this.
But, if you really want to use bool methods instead of just returning the enum, maybe s/IsUnknown/IsUsageUnknown/ or something like that?

File: data/dxl/minidump/AnySubq-With-NonScalarSubqueryChild-2.mdp [hunk 20]
```diff
@@ -607,7 +600,6 @@
                     <dxl:SortingColumnList>
                       <dxl:SortingColumn ColId="10" SortOperatorMdid="0.97.1.0" SortOperatorName="&lt;" SortNullsFirst="false"/>
                       <dxl:SortingColumn ColId="11" SortOperatorMdid="0.664.1.0" SortOperatorName="&lt;" SortNullsFirst="false"/>
-                      <dxl:SortingColumn ColId="12" SortOperatorMdid="0.97.1.0" SortOperatorName="&lt;" SortNullsFirst="false"/>
```
[the remark this answers is not in the document]
Comment (asubramanya) [reply data/dxl/minidump/AnySubq-With-NonScalarSubqueryChild-2.mdp]: This column is not being referenced anywhere and its ok to remove it. The other columns are either system columns or distribution keys or they are referenced.

File: libgpopt/src/operators/CLogicalGet.cpp (hunk 19 above)
Comment (asubramanya) [reply libgpopt/src/operators/CLogicalGet.cpp]: Sure. Its probably better to use the GetUsage() method.

File: libgpopt/src/operators/CLogical.cpp (hunk 18 above)
Comment (asubramanya) [reply libgpopt/src/operators/CLogical.cpp]: Changed to check for Unused rather than not being Used.

File: libgpopt/include/gpopt/base/CUtils.h (hunk 7 above)
Comment (asubramanya) [reply libgpopt/include/gpopt/base/CUtils.h]: Seems like this change is no longer needed now. So reverted it back to the old code.

File: libgpopt/src/base/CColumnFactory.cpp [hunk 21]
```diff
@@ -172,6 +172,7 @@ CColumnFactory::PcrCreate
 	// ensure uniqueness
 	GPOS_ASSERT(NULL == LookupColRef(id));
 	m_sht.Insert(colref);
+	colref->MarkAsUsed();
```
[the remark this answers is not in the document]
Comment (hardikar) [reply libgpopt/src/base/CColumnFactory.cpp]: this comment is wrong. You can't define an interface based on where it's called. I'd rather there be no comment.

File: libgpopt/src/translate/CTranslatorDXLToExpr.cpp [hunk 22]
```diff
@@ -58,6 +58,8 @@ using namespace gpmd;
 using namespace gpdxl;
 using namespace gpopt;
 
+//BOOL CTranslatorDXLToExpr::translating = false;
```
[the remark this answers is not in the document]
Comment (hardikar) [reply libgpopt/src/translate/CTranslatorDXLToExpr.cpp]: This still needs to be removed.

File: libgpopt/src/translate/CTranslatorDXLToExprUtils.cpp (hunk 12 above)
Comment (hardikar) [reply libgpopt/src/translate/CTranslatorDXLToExprUtils.cpp]: This still needs to be done?

Comment (hardikar) [reply libgpopt/src/translate/CTranslatorDXLToExprUtils.cpp]: I tried doing this - looks like it's going to be more work than I anticipated. It can be done later. Ignore it :)
